src/dune_engine/print_diff.ml

```diff
@@ -15,7 +15,73 @@ let resolve_link_for_git path =
     User_error.raise
       [ Pp.textf "Unable to resolve symlink %s" (Path.to_string path) ]
 
-let print ?(skip_trailing_cr = Sys.win32) annots path1 path2 =
+module Diff = struct
+  type t =
+    { loc : Loc.t option
+    ; output : string
+    }
+
+  let print { loc; output } =
+    Option.iter loc ~f:(fun loc ->
+        Loc.pp loc |> Pp.map_tags ~f:(fun Loc.Loc -> []) |> Ansi_color.print);
+    print_string output
+end
+
+type command =
+  { dir : Path.t
+  ; metadata : Process.metadata
+  ; prog : Path.t
+  ; args : string list
+  }
+
+module Job : sig
```

Comment:
The `Job` thing seems a bit too general for what we're doing here. I noticed that b/c the the type parameter is basically always `unit` or `'a`. There's also no need for a fail constructor because we know we always run the same fallback at the end - that trims the necessity for `'a`. So it seems like a pipeline that is basically `command list` should be enough:

```
val run_until_failure : command list -> 'a Fiber.t
val capture : command list -> Diff.t option Fiber.t
```

Reply:
Yes, I agree that it does not look like the right abstraction. I initially tried the `command list` approach, but couldn't make it work. In particular note that in the `Some cmd` case, the job looks like `command <|> fail` - there's no `fallback` involved in that case. So I'm not sure how to improve it.

(I also tried to make a bit more abstract and use a free monad, but it didn't make a lot of sense - as you mentioned it's only useful for `unit` return values)

Reply:
I finally managed to do it without GADTs. Let me know what you think.

Reply:
LGTM. Just some naming issue.

Instead of the non-suggestive `Job`, how about just `With_fallback` or `Command_with_fallback`?

Can you rename `<|>`? It's a bit too cute and the signature doesn't really have the signature of normal alternation. An example of a new name could be:

```
val run_with_fallback : command -> fallback:t -> t
```

Reply:
done, I went with `With_fallback.run`